File: 3. Longest Substring Without Repeating Characters/3. Longest Substring Without Repeating Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        seen = set()
        left = 0
        max_length = 0

        for i in range(len(s)):
            while s[i] in seen:
                seen.remove(s[left])
                left += 1

            seen.add(s[i])

            curr_length = i - left + 1
            max_length = max(max_length, curr_length)

        return max_length

File: 3. Longest Substring Without Repeating Characters/3. Longest Substring Without Repeating Characters.java
import java.util.HashSet;
import java.util.Set;

class Solution {

  public int lengthOfLongestSubstring(String s) {
    Set<Character> seen = new HashSet<>();
    int left = 0;
    int maxLength = 0;

    for (int i = 0; i < s.length(); i++) {
      while (seen.contains(s.charAt(i))) {
        seen.remove(s.charAt(left));
        left++;
      }

      seen.add(s.charAt(i));

      int currLength = i - left + 1;
      maxLength = Math.max(maxLength, currLength);
    }

    return maxLength;
  }
}
